up-l1/mqtt.adoc

```diff
@@ -35,8 +35,100 @@ MQTT is an OASIS standard messaging protocol for the Internet of Things (IoT). I
 
 For more information, please refer to https://mqtt.org/
 
-
 == Specifications
 
+=== UAttribute Mapping
+
+MQTT 5 supports custom header fields, and we leverage this to map certain UAttribute values into the MQTT header.
+
+[cols="1,1"]
+|===
+| MQTT Header Property Code | Value (key:value)
+| UserProperty
+| ("upriority","{UAttribute.priority}")
+| CorrelationData
+| ("reqid", "{UAttribute.reqId}")
+| UserProperty
+| ("sink","{UAttribute.sink}") # Short Form Serialized
+| UserProperty
+| ("type", "{UAttribute.type}")
+|===
+
+The upriority field is needed as mqtt's QoS is very limited, so this will allow listeners to prioritize messages.
+
+The Correlation Data Property Code contains the request correlation id.
+
+The sink field is used to correctly map a message to the correct uE as the topic name may not contain all of the information. The sink field **MUST** be a short form serialized UUri.
+
+The type field provides information on the type of message being sent which can be used by a UListener to select messages on the topic.
+
+=== URI Mapping
+
+The URI mapping follows the short form UUri format, with the addition of a location identifier to determine if the message is intended for a local or cloud destination.
+
+* Publish Message
+  - Use source UUri
+  - {location_identifier}/{UUri.UAuthority.number}/{UUri.UEntity.id}/{UUri.UEntity.version_major}/{UUri.UResource.id}
+
+* Notification, Request or Response Message
+  - Use sink UUri
+  - {location_identifier}/{UUri.UAuthority.number}/{UUri.UEntity.id}/{UUri.UEntity.version_major}/{UUri.UResource.id}
+
+Location Identifier is one of the following:
+  - d2c - This is to identify remote messages going from a device to the cloud.
+  - c2d - This is to identify remote messages going from the cloud to a device.
+  - upl - This is to identify local messages on the device.
+
+Examples:
+
+[source]
+----
+UUri {
+  authority: UAuthority {
+    name: "Vehicle_1",
+    number: 192.168.0.0,
+  },
+  entity: UEntity {
+    name: "body.access",
+    version_major: 1,
+    id: 1234,
+  }
+  resource: UResource {
+    name: "door",
+    instance: "front_left",
+    message: "Door",
+    id: 5678,
+  }
+}
+----
+
+* A Publish Message: `d2c/192.168.0.0/1234/1/5678`
```

Comment:
2 questions:
1. Why are we selecting IPs and IDs instead of the names?
2. Question I had asked in slack: How do we build topics when the UMessageType is unknown, which is true for register_listener()

Reply:
1. @lynn2009 and others I think have concerns about the topics getting too long, and this follows the short uri format for the UURIs. @stevenhartley any other thoughts?

Reply:
Oh okay all good, just wanted to understand why the change happened

Reply:
> my understanding is that the short uri format can have the uAuthority be either the ip address or an id. @stevenhartley is that correct?

Correct. So we would use ID (the VIN) for talking with the vehicle.